### PR TITLE
Externalize Responder interface

### PR DESCRIPTION
**Pull request by rkatsyuryna**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -97,9 +97,9 @@ Factory method to create a new instance of a responder server
 #### listen()
 start listening for incoming requests
 
-Class: Responder
+Interface: Responder
 ---------------------------------------------------
-Component which holds original request message and create response base on it.
+Implementation of this interface holds original request message and create response base on it.
```
Comment (vso-tc) [inline README.md]: Does a client of this class has to know about these hairy details that there's some message inside the responder? Maybe all he needs to know that there's some magic behind send and sendAck that make them work.

File: core/src/main/java/io/github/tcdl/api/Responder.java [hunk 2]
```diff
@@ -1,91 +1,28 @@
...
-import io.github.tcdl.message.MessageFactory;
-import io.github.tcdl.support.Utils;
-import org.apache.commons.lang3.Validate;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import static io.github.tcdl.api.message.Acknowledge.AcknowledgeBuilder;
 
 /**
  * {@link Responder} is a component which holds original request message and create response base on it.
```
Comment (vso-tc) [inline core/src/main/java/io/github/tcdl/api/Responder.java]: Not a valid javadoc anymore

File: core/src/main/java/io/github/tcdl/api/Responder.java [hunk 3]
```diff
@@ -1,91 +1,28 @@
...
-        ackBuilder.withResponderId(responderId);
-        ackBuilder.withTimeoutMs(timeoutMs != null && timeoutMs > -1 ? timeoutMs : null);
-        ackBuilder.withResponsesRemaining(responsesRemaining == null ? 1 : responsesRemaining);
-
-        Message message = this.messageFactory.createResponseMessage(this.messageBuilder, ackBuilder.build(), null);
-        sendMessage(message);
-
-        return message;
-    }
+    Message sendAck(Integer timeoutMs, Integer responsesRemaining);
```
Comment (vso-tc) [inline core/src/main/java/io/github/tcdl/api/Responder.java:18]: The return value here is another reason why we have Message in the api :smile:

File: core/src/main/java/io/github/tcdl/api/ResponderServer.java [hunk 4]
```diff
@@ -34,10 +34,10 @@ private ResponderServer(String namespace, MessageTemplate messageTemplate, MsbCo
     }
 
     /**
-     * Create a new instance of a {@link ResponderServer}.
+     * Create a new instance of {@link ResponderServer}.
      *
-     * @param namespace       topic on a bus for listening incoming requests
-     * @param messageTemplate template which will be used for creating response messages, object of class {@link MessageTemplate}
+     * @param namespace       topic on a bus for listening on incoming requests
+     * @param messageTemplate template used for creating response messages, instance of type {@link MessageTemplate}
```
Comment (vso-tc) [inline core/src/main/java/io/github/tcdl/api/ResponderServer.java]: The phrase that goes after comma seems to be redundant. The type is evident from method signature.

File: core/src/main/java/io/github/tcdl/api/ResponderServer.java [hunk 5]
```diff
@@ -55,7 +55,7 @@ public ResponderServer listen() {
         channelManager.subscribe(namespace,
                 message -> {
                         LOG.debug("Received message with id {} from topic {}", message.getId(), namespace);
-                        Responder responder = new Responder(messageTemplate, message, msbContext);
+                        ResponderImpl responder = new ResponderImpl(messageTemplate, message, msbContext);
```
Comment (anstr) [inline core/src/main/java/io/github/tcdl/api/ResponderServer.java:58]: Suppose we should write Responder responder = new ResponderImpl(messageTemplate, message, msbContext);

File: core/src/main/java/io/github/tcdl/api/ResponderServer.java [hunk 6]
```diff
@@ -64,19 +64,19 @@ public ResponderServer listen() {
...
+         * Execute business logic and send response.
+         * @param request request received from a bus
+         * @param responder object of type {@link Responder} which will be used for sending response
          * @throws Exception if some problems during execution business logic or sending response were occurred
          */
         void process(Payload request, Responder responder) throws Exception;
     }
 
-    protected void onResponder(Responder responder) {
+    void onResponder(ResponderImpl responder) {
```
Comment (anstr) [inline core/src/main/java/io/github/tcdl/api/ResponderServer.java:79]: Here we should use Responder,not implementation.

File: core/src/main/java/io/github/tcdl/api/ResponderServer.java [hunk 7]
```diff
@@ -88,7 +88,7 @@ protected void onResponder(Responder responder) {
         }
     }
 
-    private void errorHandler(Responder responder, Exception exception) {
+    private void errorHandler(ResponderImpl responder, Exception exception) {
```
Comment (anstr) [inline core/src/main/java/io/github/tcdl/api/ResponderServer.java:91]: The same,we should use Responder not Impl.

File: core/src/test/java/io/github/tcdl/api/ResponderServerTest.java [hunk 8]
```diff
@@ -76,7 +77,7 @@ public void testResponderServerProcessWithError() throws Exception {
 
         // simulate incoming request
         ArgumentCaptor<Payload> responseCaptor = ArgumentCaptor.forClass(Payload.class);
-        Responder responder = spy(new Responder(messageTemplate, TestUtils.createMsbRequestMessageNoPayload(TestUtils.getSimpleNamespace()), msbContext));
+        ResponderImpl responder = spy(new ResponderImpl(messageTemplate, TestUtils.createMsbRequestMessageNoPayload(TestUtils.getSimpleNamespace()), msbContext));
```
Comment (anstr) [inline core/src/test/java/io/github/tcdl/api/ResponderServerTest.java:80]: I would suggest Responder responder = spy(new ResponderImpl(messageTemplate, TestUtils.createMsbRequestMessageNoPayload(TestUtils.getSimpleNamespace()), msbContext));

File: README.md (hunk 1 above)
Comment (rkatsyuryna) [inline README.md]: well i see no problem to have this in readme file. More over it was there before

File: core/src/main/java/io/github/tcdl/api/Responder.java (hunk 3 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/api/Responder.java:18]: actually the only why we have it is - tests(

File: core/src/main/java/io/github/tcdl/api/ResponderServer.java (hunk 4 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/api/ResponderServer.java]: agree

File: core/src/main/java/io/github/tcdl/api/ResponderServer.java (hunk 5 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/api/ResponderServer.java:58]: this is internal logic - private methods, and also we make use of ResponderImpl.getOriginalMessage which is not part of Responder contract

File: core/src/test/java/io/github/tcdl/api/ResponderServerTest.java (hunk 8 above)
Comment (rkatsyuryna) [inline core/src/test/java/io/github/tcdl/api/ResponderServerTest.java:80]: we cant as i need to pass ResponderImpl to package private method onResponder()

File: core/src/main/java/io/github/tcdl/api/Responder.java (hunk 2 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/api/Responder.java]: agree

**Pull request merged**